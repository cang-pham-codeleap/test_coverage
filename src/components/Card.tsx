import React from "react";

/**
 * Props for the Card component
 * @property {React.ReactNode} children - The content to be displayed within the card
 * @property {string} [className] - Additional CSS classes
 * @property {string} [title] - Optional card title
 * @property {React.ReactNode} [footer] - Optional footer content
 * @property {'default' | 'bordered' | 'elevated'} [variant='default'] - Visual style of the card
 * @property {'none' | 'sm' | 'md' | 'lg'} [padding='md'] - Padding size inside the card
 * @property {boolean} [noDivider=false] - Whether to hide the divider between title, content, and footer
 */
export interface CardProps {
  children: React.ReactNode;
  className?: string;
  title?: string;
  footer?: React.ReactNode;
  variant?: "default" | "bordered" | "elevated";
  padding?: "none" | "sm" | "md" | "lg";
  noDivider?: boolean;
  [key: string]: any; // For additional HTML attributes
}

/**
 * Card component for displaying content in a contained box
 *
 * @example
 * ```tsx
 * <Card>Basic card content</Card>
 *
 * <Card
 *   title="Card Title"
 *   variant="elevated"
 *   footer={<Button>Action</Button>}
 * >
 *   Card with title and footer
 * </Card>
 *
 * <Card padding="lg" variant="bordered">
 *   Card with large padding and border
 * </Card>
 * ```
 */
export const Card: React.FC<CardProps> = ({
  children,
  className = "",
  title,
  footer,
  variant = "default",
  padding = "md",
  noDivider = false,
  ...props
}) => {
  // Variant classes
  const variantClasses = {
    default: "bg-white",
    bordered: "bg-white border border-gray-200",
    elevated: "bg-white shadow-md",
  };

  // Padding classes
  const paddingClasses = {
    none: "p-0",
    sm: "p-3",
    md: "p-4",
    lg: "p-6",
  };

  // Title and footer padding - smaller padding on the sides when content has none
  const headerFooterPadding =
    padding === "none"
      ? "px-4 py-3"
      : `py-3 ${paddingClasses[padding].replace("p-", "px-")}`;

  // Create CSS classes based on props
  const cardClasses = [
    "rounded-lg overflow-hidden",
    variantClasses[variant],
    className,
  ]
    .filter(Boolean)
    .join(" ");

  // Content classes
  const contentClasses = paddingClasses[padding];

  // Divider class
  const dividerClass = "border-t border-gray-200";

  return (
    <div className={cardClasses} {...props}>
      {title && (
        <div className={`${headerFooterPadding} font-medium`}>{title}</div>
      )}

      {title && !noDivider && <div className={dividerClass} />}

      <div className={contentClasses}>{children}</div>

      {footer && !noDivider && <div className={dividerClass} />}

      {footer && <div className={headerFooterPadding}>{footer}</div>}
    </div>
  );
};

export default Card;
