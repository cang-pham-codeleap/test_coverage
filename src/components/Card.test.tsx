/**
 * Card component tests
 * - Tests basic rendering
 * - Tests different variants and padding sizes
 * - Tests title and footer rendering
 * - Tests divider visibility
 * - Tests custom styling
 */

import React from "react";
import { render, screen } from "@testing-library/react";
import Card from "./Card";

describe("Card Component", () => {
  // Test basic rendering
  test("renders card with children", () => {
    render(<Card>Card Content</Card>);
    const cardContent = screen.getByText("Card Content");
    expect(cardContent).toMatchSnapshot();
  });

  // Test default variant
  test("renders default variant by default", () => {
    render(<Card>Default Card</Card>);
    // Find card by its content and then get the parent div (card container)
    const cardContent = screen.getByText("Default Card");
    const cardElement = cardContent.parentElement?.parentElement;
    expect(cardElement?.className).toMatchSnapshot();
  });

  // Test bordered variant
  test("renders bordered variant when specified", () => {
    render(<Card variant="bordered">Bordered Card</Card>);
    const cardContent = screen.getByText("Bordered Card");
    const cardElement = cardContent.parentElement?.parentElement;
    expect(cardElement?.className).toMatchSnapshot();
  });

  // Test elevated variant
  test("renders elevated variant when specified", () => {
    render(<Card variant="elevated">Elevated Card</Card>);
    const cardContent = screen.getByText("Elevated Card");
    const cardElement = cardContent.parentElement?.parentElement;
    expect(cardElement?.className).toMatchSnapshot();
  });

  // Test different padding sizes
  test("applies correct padding size classes", () => {
    const { rerender } = render(<Card padding="none">No Padding</Card>);
    let cardContent = screen.getByText("No Padding");
    expect(cardContent.className).toMatchSnapshot();

    rerender(<Card padding="sm">Small Padding</Card>);
    cardContent = screen.getByText("Small Padding");
    expect(cardContent.className).toMatchSnapshot();

    rerender(<Card padding="md">Medium Padding</Card>);
    cardContent = screen.getByText("Medium Padding");
    expect(cardContent.className).toMatchSnapshot();

    rerender(<Card padding="lg">Large Padding</Card>);
    cardContent = screen.getByText("Large Padding");
    expect(cardContent.className).toMatchSnapshot();
  });

  // Test title rendering
  test("renders title when provided", () => {
    const title = "Card Title";
    render(<Card title={title}>Card with Title</Card>);
    const titleElement = screen.getByText(title);
    expect(titleElement).toMatchSnapshot();
  });

  // Test footer rendering
  test("renders footer when provided", () => {
    const footer = "Card Footer";
    render(<Card footer={<div>{footer}</div>}>Card with Footer</Card>);
    const footerElement = screen.getByText(footer);
    expect(footerElement).toMatchSnapshot();
  });

  // Test divider visibility
  test("renders dividers by default when title or footer is present", () => {
    render(
      <Card title="Card Title" footer={<div>Footer</div>}>
        Card Content
      </Card>
    );
    // Using querySelectorAll to find dividers (elements with border-t class)
    // We need to use querySelector on the container since querySelectorAll doesn't exist on RTL's screen
    const container =
      screen.getByText("Card Content").parentElement?.parentElement;
    const dividers = container?.querySelectorAll(".border-t");
    expect(dividers?.length).toBe(2); // One for title and one for footer
  });

  // Test hiding dividers
  test("doesn't render dividers when noDivider is true", () => {
    render(
      <Card title="Card Title" footer={<div>Footer</div>} noDivider>
        Card Content
      </Card>
    );
    const container =
      screen.getByText("Card Content").parentElement?.parentElement;
    const dividers = container?.querySelectorAll(".border-t");
    expect(dividers?.length).toBe(0);
  });

  // Test custom class name
  test("applies custom className when provided", () => {
    const customClass = "custom-card-class";
    render(<Card className={customClass}>Custom Card</Card>);
    const cardContent = screen.getByText("Custom Card");
    const cardElement = cardContent.parentElement?.parentElement;
    expect(cardElement?.className).toMatchSnapshot();
  });

  // Test additional props
  test("applies additional props to the card element", () => {
    render(
      <Card data-testid="custom-card" id="special-card">
        Card with Custom Props
      </Card>
    );
    const cardElement = screen.getByTestId("custom-card");
    expect(cardElement).toMatchSnapshot();
  });

  // Test combining complex features
  test("correctly renders a complex card with all features", () => {
    render(
      <Card
        title="Complex Card"
        variant="bordered"
        padding="lg"
        footer={<button>Action Button</button>}
      >
        Complex Card Content
      </Card>
    );

    expect(screen.getByText("Complex Card")).toMatchSnapshot();
    expect(screen.getByText("Complex Card Content")).toMatchSnapshot();
    expect(screen.getByRole("button")).toMatchSnapshot();

    const cardContent = screen.getByText("Complex Card Content");
    const cardElement = cardContent.parentElement?.parentElement;
    expect(cardElement).toMatchSnapshot();
  });
});
