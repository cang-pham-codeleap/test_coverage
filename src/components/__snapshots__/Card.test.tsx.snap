// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Card Component applies additional props to the card element 1`] = `
<div
  class="rounded-lg overflow-hidden bg-white"
  data-testid="custom-card"
  id="special-card"
>
  <div
    class="p-4"
  >
    Card with Custom Props
  </div>
</div>
`;

exports[`Card Component applies correct padding size classes 1`] = `"p-0"`;

exports[`Card Component applies correct padding size classes 2`] = `"p-3"`;

exports[`Card Component applies correct padding size classes 3`] = `"p-4"`;

exports[`Card Component applies correct padding size classes 4`] = `"p-6"`;

exports[`Card Component applies custom className when provided 1`] = `""`;

exports[`Card Component correctly renders a complex card with all features 1`] = `
<div
  class="py-3 px-6 font-medium"
>
  Complex Card
</div>
`;

exports[`Card Component correctly renders a complex card with all features 2`] = `
<div
  class="p-6"
>
  Complex Card Content
</div>
`;

exports[`Card Component correctly renders a complex card with all features 3`] = `
<button>
  Action Button
</button>
`;

exports[`Card Component correctly renders a complex card with all features 4`] = `
<div>
  <div
    class="rounded-lg overflow-hidden bg-white border border-gray-200"
  >
    <div
      class="py-3 px-6 font-medium"
    >
      Complex Card
    </div>
    <div
      class="border-t border-gray-200"
    />
    <div
      class="p-6"
    >
      Complex Card Content
    </div>
    <div
      class="border-t border-gray-200"
    />
    <div
      class="py-3 px-6"
    >
      <button>
        Action Button
      </button>
    </div>
  </div>
</div>
`;

exports[`Card Component renders bordered variant when specified 1`] = `""`;

exports[`Card Component renders card with children 1`] = `
<div
  class="p-4"
>
  Card Content
</div>
`;

exports[`Card Component renders default variant by default 1`] = `""`;

exports[`Card Component renders elevated variant when specified 1`] = `""`;

exports[`Card Component renders footer when provided 1`] = `
<div>
  Card Footer
</div>
`;

exports[`Card Component renders title when provided 1`] = `
<div
  class="py-3 px-4 font-medium"
>
  Card Title
</div>
`;
